subsys/bluetooth/host/gatt.c

```diff
@@ -711,9 +704,13 @@ static ssize_t db_hash_read(struct bt_conn *conn,
 	/* Check if db_hash is already pending in which case it shall be
 	 * generated immediately instead of waiting for the work to complete.
 	 */
-	err = k_delayed_work_cancel(&db_hash_work);
-	if (!err) {
-		db_hash_gen(true);
+	err = k_work_delayable_busy_get(&db_hash_work);
```

Comment:
The idiom of checking for a delay and then cancelling is unnecessarily complex; see the "Don't over optimize" part of the Best Practices section in the issue description in #33104.

Also the solution here is incorrect due to the bug in #34351.   When that's merged, this would be a preferred approach instead of the entire check-busy/cancel/compute loop.

    if (!k_work_cancel_delayable_sync(&db_hash_work, &sync)) {
        db_hash_gen(true);
    }

Another reason to do that is that the `k_work_sync` object cannot be allocated on the stack in some architectures; see the function call documentation.  Better to create `db_hash_sync` as a file-scope object sibling to `db_hash_work`, and use that.

Reply:
@joerchan The above comment was incorrect in several ways and has been updated.

Reply:
@pabigot I've updated this, the DB_HASH_VALID flag should prevent a double calculation when it has to be canceled.
I didn't want to use the flush as I'd like to not block the RX thread. The cancel sync should handle that they don't both end up calculating at the same time.